burstphoto/align.swift

```diff
@@ -165,44 +617,102 @@ func avg_pool(_ input_texture: MTLTexture, _ scale: Int) -> MTLTexture {
-    output_texture_descriptor.usage = [.shaderRead, .shaderWrite]
-    let output_texture = device.makeTexture(descriptor: output_texture_descriptor)!
+    // create a temp texture blurred along x-axis only and the output texture, blurred along both x- and y-axis
+    let blur_x = texture_like(in_texture)
+    let blur_xy = texture_like(in_texture)
     
-    // run the metal kernel
+    // blur the texture along the x-axis
     let command_buffer = command_queue.makeCommandBuffer()!
     let command_encoder = command_buffer.makeComputeCommandEncoder()!
-    let state = average_texture_sums_state
+    let state = blur_mosaic_x_state
     command_encoder.setComputePipelineState(state)
-    let threads_per_grid = MTLSize(width: output_texture.width, height: output_texture.height, depth: 1)
+    let threads_per_grid = MTLSize(width: in_texture.width, height: in_texture.height, depth: 1)
     let threads_per_thread_group = get_threads_per_thread_group(state, threads_per_grid)
-    command_encoder.setTexture(input_texture, index: 0)
-    command_encoder.setTexture(output_texture, index: 1)
-    command_encoder.setBytes([Int32(n)], length: MemoryLayout<Int32>.stride, index: 0)
+    command_encoder.setTexture(in_texture, index: 0)
+    command_encoder.setTexture(blur_x, index: 1)
+    command_encoder.setBytes([Int32(kernel_size)], length: MemoryLayout<Int32>.stride, index: 0)
+    command_encoder.setBytes([Int32(mosaic_pattern_width)], length: MemoryLayout<Int32>.stride, index: 1)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+
+    // blur the texture along the y-axis
+    let state2 = blur_mosaic_y_state
+    command_encoder.setComputePipelineState(state2)
+    command_encoder.setTexture(blur_x, index: 0)
+    command_encoder.setTexture(blur_xy, index: 1)
     command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    
     command_encoder.endEncoding()
     command_buffer.commit()
-
-    return output_texture
+    
+    return blur_xy
 }
 
 
-func build_pyramid(_ input_texture: MTLTexture, _ downscale_factor_list: Array<Int>) -> Array<MTLTexture> {
+func extend_texture(_ in_texture: MTLTexture, _ pad_left: Int, _ pad_right: Int, _ pad_top: Int, _ pad_bottom: Int) -> MTLTexture {
     
-    // iteratively resize the current layer in the pyramid
-    var pyramid: Array<MTLTexture> = []
-    for (i, downscale_factor) in downscale_factor_list.enumerated() {
-        if i == 0 {
-            pyramid.append(avg_pool(input_texture, downscale_factor))
-        } else {
-            pyramid.append(avg_pool(pyramid.last!, downscale_factor))
-        }
-        // print("layer.shape: ", pyramid.last!.width, pyramid.last!.height)
-    }
-    return pyramid
+    let out_texture_descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .r16Float, width: in_texture.width+pad_left+pad_right, height: in_texture.height+pad_top+pad_bottom, mipmapped: false)
```

Comment:
It's unclear why some of these are using 16bit floats while others are using 32bits

Reply:
All textures storing input images can be 16 bit as this is the maximum information we are getting from the DNGs. As soon as calculations are performed on the values (besides addition and subtraction), the app uses 32 bit textures for improved fidelity. The complete alignment requires only 16 bit textures as pixel values are only shifted. The merging requires 32 bit textures as more complex calculations are performed.